api/src/main/java/ai/djl/engine/Engine.java

```diff
@@ -306,6 +307,14 @@ public int getGpuCount() {
      */
     public abstract GradientCollector newGradientCollector();
 
+    /**
+     * Returns a new instance of {@link GradientCollector}.
+     *
+     * @param block the block that the gradient collector attaches
+     * @return a new instance of {@link GradientCollector}
+     */
+    public abstract GradientCollector newGradientCollector(Block block);
+
     /**
      * Returns a new instance of {@link ParameterServer}.
```

Comment:
Can you also update `trainer.newGradientCollector()` to use this? You can change it's definition to:

```java
    public GradientCollector newGradientCollector() {
        return manager.getEngine().newGradientCollector(getModel().getBlock());
    }
```

Reply:
The tricky part is that inside the newGradientCollector(), the zeroGradient is not called, while inside newGradientCollector(Block block), it is. This is to avoid the efficiency and memory leak issue with global search scope as discussed  last week.

Reply:
Another difference is 
```java
        boolean wasCollecting = isCollecting.getAndSet(true);
        if (wasCollecting) {
            throw new IllegalStateException(
                    "A PtGradientCollector is already collecting. Only one can be collecting at a"
                            + " time");
        }
```
is not called inside `PtGradientCollector(Block block)`.

Reply:
In `zeroGradients` `array.getGradient().subi(array.getGradient());` is still used. 
Couldn't it be replaced with `JniUtils.zeroGrad((PtNDArray) array);` as used in `zeroGradients(Block block)`?

Reply:
> As @zachgk mentioned ... in case you go with the approach it must also work when using a trainer.

That would be ideal. But it has the efficient and memory issue for now. So we should disable it for now.

> Are you sure that you get all relevant NDArrays with Gradient by getting the parameter NDArrays from block?

The gradientCollector is specified to a block. By defination it does search only within a block.